### ISSUE TITLE
[Policy Assistant] Summary Table for (B)ANP and NetPol

### ISSUE THREAD

**Issue by huntergregory** [kubernetes-sigs/network-policy-api#153]
Requirement for #150.

Dependency: #152

Repurposing `cyclonus analyze --mode=explain` (see `ParseMode` in _analyze.go_). This function is also used in running the conformance tests via `cyclonus generate`. The function displays a summary table of all ingress rules for the set of NetworkPolicies, then all egress rules for the set.

Current Output:
```
$ ./cyclonus analyze --mode=explain --use-example-policies
INFO[2023-10-10T00:12:51-07:00] log level set to 'info'
explained policies:
+---------+--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
|  TYPE   |       TARGET       |                            SOURCE RULES                             |           PEER           |       PORT/PROTOCOL       |
+---------+--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
| Ingress | namespace: default | default/accidental-and                                              | namespace: default       | all ports, all protocols  |
|         | Match labels:      | default/accidental-or                                               | pods: Match labels:      |                           |
|         |   a: b             |                                                                     |   role: client           |                           |
+         +                    +                                                                     +--------------------------+                           +
|         |                    |                                                                     | namespace: Match labels: |                           |
|         |                    |                                                                     |   user: alice            |                           |
|         |                    |                                                                     | pods: Match labels:      |                           |
|         |                    |                                                                     |   role: client           |                           |
+         +                    +                                                                     +--------------------------+                           +
|         |                    |                                                                     | namespace: Match labels: |                           |
|         |                    |                                                                     |   user: alice            |                           |
|         |                    |                                                                     | pods: all                |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
|         | namespace: default | default/allow-nothing-to-v2-all-web                                 | no pods, no ips          | no ports, no protocols    |
|         | Match labels:      |                                                                     |                          |                           |
|         |   all: web         |                                                                     |                          |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
|         | namespace: default | default/allow-specific-port-from-role-monitoring-to-app-apiserver   | namespace: default       | port 5000 on protocol TCP |
|         | Match labels:      |                                                                     | pods: Match labels:      |                           |
|         |   app: apiserver   |                                                                     |   role: monitoring       |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
|         | namespace: default | default/allow-from-app-bookstore-to-app-bookstore-role-api          | namespace: default       | all ports, all protocols  |
|         | Match labels:      |                                                                     | pods: Match labels:      |                           |
|         |   app: bookstore   |                                                                     |   app: bookstore         |                           |
|         |   role: api        |                                                                     |                          |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+                           +
|         | namespace: default | default/allow-from-multiple-to-app-bookstore-role-db                | namespace: default       |                           |
|         | Match labels:      |                                                                     | pods: Match labels:      |                           |
|         |   app: bookstore   |                                                                     |   app: bookstore         |                           |
|         |   role: db         |                                                                     |   role: api              |                           |
+         +                    +                                                                     +--------------------------+                           +
|         |                    |                                                                     | namespace: default       |                           |
|         |                    |                                                                     | pods: Match labels:      |                           |
|         |                    |                                                                     |   app: bookstore         |                           |
|         |                    |                                                                     |   role: search           |                           |
+         +                    +                                                                     +--------------------------+                           +
|         |                    |                                                                     | namespace: default       |                           |
|         |                    |                                                                     | pods: Match labels:      |                           |
|         |                    |                                                                     |   app: inventory         |                           |
|         |                    |                                                                     |   role: web              |                           |
+         +--------------------+---------------------------------------------------------------------+--------------------------+---------------------------+
```

_EDIT: below images are slightly outdated. See new images in further comments._

We can repurpose this for ANP (highlighting changes):
![image](https://github.com/kubernetes-sigs/network-policy-api/assets/42728408/3a986570-aebf-4d3b-8a64-ac2f7f5e00e1)

Example with multiple Network Policies:
![image](https://github.com/kubernetes-sigs/network-policy-api/assets/42728408/ae977926-bad6-4f31-8fe9-79bdf8773b2a)

**Comment by huntergregory** [comment]
Revised idea for tables:
![image](https://github.com/kubernetes-sigs/network-policy-api/assets/42728408/f7ef9e27-c767-4df3-9689-8225de05e13f)

**Comment by Peac36** [comment]
Hey @huntergregory, I would like to work on this and I believe I know what is expected here however I have two questions:

* Should we print priority info if the peer is NetworkPolicyV1?
* Are there examples (B)ANP similar to the NetworkPolicy that I can use for testing?

**Comment by huntergregory** [comment]
Thanks for the help with this issue @Peac36 !

For testing, there are some policies in _test/integration/integration_test.go_ which might be helpful.

NetworkPolicyV1 doesn't quite have a priority concept. Instead, a NPv1 defines a target (based on its pod selector and the namespace it resides in) and has a list of allowed peers (e.g. for ingress, traffic goes from a peer to a target). For an isolated Pod, if any NPv1 allows ingress to the Pod, then ingress is allowed. There is similar logic for egress. [This section](https://kubernetes.io/docs/concepts/services-networking/network-policies/#the-two-sorts-of-pod-isolation) of the documentation explains this further.

So we'll have to be creative about the action/priority column for NPv1 (I was thinking to just say `NPv1: peers allowed` in this column).

Happy to help with any other questions/clarifications 🙂

**Comment by Peac36** [comment]
Thanks @huntergregory, I've just opened the PR for this. Let's continue the discussion there.

**Comment by huntergregory** [comment]
Fixed in #188
